pkg/cli/build/validate.go

```diff
@@ -0,0 +1,102 @@
+package build
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"slices"
+	"strings"
+	"time"
+
+	"github.com/suse-edge/edge-image-builder/pkg/cli/cmd"
+	"github.com/suse-edge/edge-image-builder/pkg/image"
+	"github.com/suse-edge/edge-image-builder/pkg/image/validation"
+	"github.com/suse-edge/edge-image-builder/pkg/log"
+	"github.com/urfave/cli/v2"
+)
+
+const (
+	checkValidationLogMessage = "Please check the log file under the validation directory for more information."
+)
+
+func Validate(_ *cli.Context) error {
+	args := &cmd.BuildArgs
+
+	validationDir := filepath.Join(args.ConfigDir, "_validation")
+	if err := os.MkdirAll(validationDir, os.ModePerm); err != nil {
+		log.Auditf("The validation directory could not be setup under the configuration directory '%s'.", args.ConfigDir)
+		return err
+	}
+
+	// This needs to occur as early as possible so that the subsequent calls can use the log
+	timestamp := time.Now().Format("Jan02_15-04-05")
+	logFilename := filepath.Join(validationDir, fmt.Sprintf("eib-validate-%s.log", timestamp))
+	log.ConfigureGlobalLogger(logFilename)
+
+	log.AuditInfo("Checking image config dir...")
+
+	if err := imageConfigDirExists(args.ConfigDir); err != nil {
+		cmd.LogError(err, checkValidationLogMessage)
+		os.Exit(1)
+	}
+
+	log.AuditInfo("Parsing image definition...")
+
+	imageDefinition, err := parseImageDefinition(args.ConfigDir, args.DefinitionFile)
+	if err != nil {
+		cmd.LogError(err, checkValidationLogMessage)
+		os.Exit(1)
+	}
+
+	ctx := &image.Context{
+		ImageConfigDir:  args.ConfigDir,
+		ImageDefinition: imageDefinition,
+	}
+
+	log.AuditInfo("Validating image definition...")
+
+	if err = validateImageDefinition(ctx); err != nil {
+		cmd.LogError(err, checkValidationLogMessage)
+		os.Exit(1)
+	}
+
+	log.AuditInfo("The specified image definition is valid.")
+
+	return nil
+}
+
+func validateImageDefinition(ctx *image.Context) *cmd.Error {
+	failedValidations := validation.ValidateDefinition(ctx)
+	if len(failedValidations) == 0 {
+		return nil
+	}
+
+	logMessageBuilder := strings.Builder{}
+	userMessageBuilder := strings.Builder{}
+
+	userMessageBuilder.WriteString("Image definition validation found the following errors:\n")
+	logMessageBuilder.WriteString("Image definition validation failures:\n")
+
+	orderedComponentNames := make([]string, 0, len(failedValidations))
+	for c := range failedValidations {
+		orderedComponentNames = append(orderedComponentNames, c)
+	}
+	slices.Sort(orderedComponentNames)
+
+	for _, componentName := range orderedComponentNames {
+		userMessageBuilder.WriteString("\t" + componentName + "\n")
+
+		for _, cf := range failedValidations[componentName] {
+			userMessageBuilder.WriteString("\t\t" + cf.UserMessage + "\n")
```

Comment:
Personally, I'm a much bigger fan of 2 spaces instead of tabs (so 4 spaces in this instance); the depth to which tabs indent looks awkward to me. How strongly do you feel about moving to tabs instead?

I don't know if there's a standard per se, but anecdotally looking around 2 spaces seems to be more common:

```
$ ln --help
Usage: ln [OPTION]... [-T] TARGET LINK_NAME
  or:  ln [OPTION]... TARGET
  or:  ln [OPTION]... TARGET... DIRECTORY
  or:  ln [OPTION]... -t DIRECTORY TARGET...
In the 1st form, create a link to TARGET with the name LINK_NAME.

[SNIP]

Mandatory arguments to long options are mandatory for short options too.
      --backup[=CONTROL]      make a backup of each existing destination file
  -b                          like --backup but does not accept an argument
  -d, -F, --directory         allow the superuser to attempt to hard link
                                directories (note: will probably fail due to
                                system restrictions, even for the superuser)
```

```
$ podman -h
Manage pods, containers and images

Usage:
  podman [options] [command]

Available Commands:
  attach      Attach to a running container
  auto-update Auto update containers according to their auto-update policy
```

```
$ git status
On branch debugging
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   docs/debugging.md
```